libcontainer/specconv/spec_linux.go

```diff
@@ -551,6 +561,21 @@ func createLibcontainerMount(cwd string, m specs.Mount) (*configs.Mount, error)
 	mnt.UIDMappings = toConfigIDMap(m.UIDMappings)
 	mnt.GIDMappings = toConfigIDMap(m.GIDMappings)
 
+	if mnt.UIDMappings == nil && mnt.GIDMappings == nil {
+		// If the mount has ridmap or idmap specified, but no idmappings, we
+		// have to copy the id mappings of the userns configuration. If the
+		// container doesn't have a userns, the config validator will error
+		// this out.
+		if mnt.Extensions&configs.EXT_ANY_IDMAP != 0 && spec.Linux != nil {
+			mnt.UIDMappings = toConfigIDMap(spec.Linux.UIDMappings)
```

Comment:
Maybe `user` namespace implementation in `runc` is always wrong? Because we may ignore a situation which is joining an existing user namespace. for example:
```js
{

          "type": "user",
          "path": "/proc/1/ns/user"
}
```
Please see:
https://github.com/opencontainers/runc/blob/32d433cf5284153dbf90af007f979f37592c50d6/libcontainer/container_linux.go#L1198-L1202
If we are joining an existing user ns, the value of `spec.Linux.*IDMappings` is required but will be ignored, so I think this logic is error.

If I'm right, I think you should read these mappings from the existing user ns path, and open an new issue to fix this user namespace implementation error.

Reply:
Ah, you're right -- this logic is wrong. In that case we could just open the nspath and use it as the userns fd for `mount_setattr(2)`. I will have to re-think how the `state.json` stuff works though...

Reply:
I have read documents about user ns, maybe user ns is a special ns, we should refuse to join an existing user ns in runc.
‘’’
It is not permitted to use setns() to reenter the caller's
              current user namespace.  This prevents a caller that has
              dropped capabilities from regaining those capabilities via
              a call to setns().
‘’’
From https://man7.org/linux/man-pages/man2/setns.2.html

Reply:
Issue opened: #4122

Reply:
Yeah, this is purely a bug in the validator and (after writing a test for this case with `idmap`, it's clear we don't have any integration tests validating that joining paths works).